### ISSUE TITLE
Server fails to listen while running SpineOpt

### ISSUE THREAD

**Issue by DillonJ** [Spine-project/Spine-Toolbox#1345]
I'm working with @marcjamali to install latest Toolbox and SpineOpt. We went the repo cloning route. When trying to run the A5 case study he is getting a sockets error as posted below.

Initially we thought it was because he was running on a virtual machine but he since tried on a different physical machine and got the very same error. I have tried running the case study and this error does not occur for me. 

Some trouble shooting steps that we've tried:
 - Tried Julia 1.6 and Julia 1.2
 - Disabled firewall
 - Commented in and out localhost entries in local hosts file
 
Anyone any suggestions? We have hit a brick wall with this one

```julia
Running SpineOpt for http://localhost:50955...
ERROR: LoadError: IOError: connect: connection refused (ECONNREFUSED)
Stacktrace:
[1] wait_connected(x::Sockets.TCPSocket)
@ Sockets
C:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.6\Sockets\src\Sockets.jl:532
[2] connect
@
C:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.6\Sockets\src\Sockets.jl:567
[inlined]
[3] connect(host::SubString{String}, port::Int64) @ Sockets
C:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.6\Sockets\src\Sockets.jl:552
[4] _communicate(::URIs.URI, ::String, ::String, ::Vararg{String, N} where N) @ SpineInterface c:\ER\Spine\SpineInterface\src\util.jl:576
[5] _get_data
@ c:\ER\Spine\SpineInterface\src\using_spinedb.jl:206 [inlined] [6] using_spinedb(db::URIs.URI, mod::Module) @ SpineInterface c:\ER\Spine\SpineInterface\src\using_spinedb.jl:253
[7] using_spinedb(url::String, mod::Module; upgrade::Bool) @ SpineInterface c:\ER\Spine\SpineInterface\src\using_spinedb.jl:245
[8] macro expansion
@ c:\ER\Spine\SpineOpt\src\run_spineopt.jl:57 [inlined] [9] macro expansion @ .\timing.jl:210 [inlined] [10] macro expansion @ c:\ER\Spine\SpineOpt\src\util\misc.jl:88 [inlined] [11] run_spineopt(url_in::String, url_out::String; upgrade::Bool, mip_solver::MathOptInterface.OptimizerWithAttributes,
lp_solver::MathOptInterface.OptimizerWithAttributes, cleanup::Bool, add_constraints::Function, update_constraints::Function, log_level::Int64, optimize::Bool, use_direct_model::Bool) @ SpineOpt c:\ER\Spine\SpineOpt\src\run_spineopt.jl:56
[12] run_spineopt(url_in::String, url_out::String) @ SpineOpt c:\ER\Spine\SpineOpt\src\run_spineopt.jl:55
[13] top-level scope
@ C:\ER\Projects\spine-cs-a5-main\run_spineopt.jl:3
in expression starting at C:\ER\Projects\spine-cs-a5-main\run_spineopt.jl:3
Initializing data structure from db...
[26-03-2021 10:49:42]     Unknown return code (1)
[26-03-2021 10:49:43] *** Archiving output files to results directory ***
[26-03-2021 10:49:43]     No output files defined for this Tool specification. Tip
[26-03-2021 10:49:43] Executing Tool SpineOpt failed
```

**Comment by manuelma** [comment]
We were afraid this could happen. Nobody else has encountered this error though. I have tried in a windows virtual machine and it's flawless. How can we investigate it? Can you provide details about the platform? Anything suspicious that might be affecting the  socket communication? Does the python socket example work in that machine? https://docs.python.org/3/howto/sockets.html

**Comment by manuelma** [comment]
[echo_server_client.zip](https://github.com/Spine-project/Spine-Toolbox/files/6219916/echo_server_client.zip)

I'm attaching a zip files with two files. Please extract those in any location, then

1. Open a windows terminal on that location and run `python echo_server.py`
2. Open a second windows terminal on that same location and run `python echo_client.py`

Is there socket communication?

**Comment by DillonJ** [comment]
Thanks @manuelma we will try that. 

I know it seems strange and I've never encountered the issue either - seemingly on the same platform.

**Comment by DillonJ** [comment]
I wonder could this be related :
https://stackoverflow.com/questions/60326237/julia-sockets-tcp-connection-refused-when-running-script-not-repl

The issue there is that the connection is attempted before the server has time to listen

**Comment by manuelma** [comment]
Thanks @DillonJ I look forward to hear about this.

**Comment by manuelma** [comment]
> I wonder could this be related :

This is certainly a possibility but it's mostly guessing at this point. Running the example I attached without failures means we 100% can do something about it. Failing to run that simple example means the simplest python socket server cannot run on that machine, because of some mysterious setting.

And in the latter case, we need to consider what to do in that situation. I guess Spine Toolbox should be able to automatically check if sockets can work in the host machine, and at that point decide: whether to advertise the server url, or just the naked db url in which case PyCall will have to do it's job. This means the user needs to configure PyCall but that's better than not being able to run SpineOpt at all!

**Comment by marcjamali** [comment]
> [echo_server_client.zip](https://github.com/Spine-project/Spine-Toolbox/files/6219916/echo_server_client.zip)
> 
> I'm attaching a zip file with two files. Please extract those in any location, then
> 
> 1. Open a windows terminal on that location and run `python echo_server.py`
> 2. Open a second windows terminal on that same location and run `python echo_client.py`
> 
> Is there socket communication?

Hello everyone,

I'm glad to help you with this project.

I did the "echo server client" test mentioned above and it seems to work.  Here are the results:

****** Server terminal ********
PS C:\Spine\echo_server_client> python echo_server.py
starting up on localhost port 10000
waiting for a connection
connection from ('127.0.0.1', 49798)
received b'This is the mess'
sending data back to the client
received b'age.  It will be'
sending data back to the client
received b' repeated.'
sending data back to the client
received b''
no data from ('127.0.0.1', 49798)
waiting for a connection 

****** Client terminal ********
PS C:\Spine\echo_server_client> python echo_client.py
connecting to localhost port 10000
sending b'This is the message.  It will be repeated.'
received b'This is the mess'
received b'age.  It will be'
received b' repeated.'
closing socket
PS C:\Spine\echo_server_client> 

Marc

**Comment by manuelma** [comment]
Hi @marcjamali 

Thanks very much, this is invaluable help!

I'm starting to think @DillonJ is onto something, and the problem is due to the server not starting. Let me ask you this, do you get any additional error messages from the python console where you launched spine toolbox from?

**Comment by marcjamali** [comment]
> Hi @marcjamali
> 
> Thanks very much, this is invaluable help!
> 
> I'm starting to think @DillonJ is onto something, and the problem is due to the server not starting. Let me ask you this, do you get any additional error messages from the python console where you launched spine toolbox from?

Hello,

No, I have nothing showing up in the console (I run it from the Windows PowerShell console with Administrator rights).  I also tried running it from the python console and nothing happens there as well.

**Comment by manuelma** [comment]
Thanks @marcjamali I will keep investigating.

I also edited the title as this is not a sockets error. The sockets work, the problem comes after the socket has been established, when the client tries to say something and the server, for some reason, is no longer (or not yet) listening.

**Comment by soininen** [comment]
Could this have anything to do with a firewall?

**Comment by manuelma** [comment]
They say they disabled the firewall. The error message they get is 'connection refused', which means there was 'communication'. The client was able to open the socket but found nothing at the other end. I think it's related to threading, not to networking. Then again is puzzling because it doesn't happen to anybody else...

**Comment by DillonJ** [comment]
> Could this have anything to do with a firewall?
In this issue I posted earlier - a similar error is caused by the server not being established when the connection is attempted - introducing a trivial delay between listen and connect solved the issue in that case - could we perhaps try to create a julia/python script to test this scenario?

https://stackoverflow.com/questions/60326237/julia-sockets-tcp-connection-refused-when-running-script-not-repl

**Comment by DillonJ** [comment]
> They say they disabled the firewall. The error message they get is 'connection refused', which means there was 'communication'. The client was able to open the socket but found nothing at the other end. I think it's related to threading, not to networking. Then again is puzzling because it doesn't happen to anybody else...

What can we do to troubleshoot further? It's a pity that @marcjamali is completely dead in the water. Would his system specs help?

**Comment by manuelma** [comment]
> could we perhaps try to create a julia/python script to test this scenario?

You can try that easily, just add `sleep(1)` before `run_spineopt(...)` in the tool spec main program file.

**Comment by DillonJ** [comment]
> > could we perhaps try to create a julia/python script to test this scenario?
> 
> You can try that easily, just add `sleep(1)` before `run_spineopt(...)` in the tool spec main program file.

Ok, we can try that... but what about the threading line of inquiry - how do we test that?

**Comment by manuelma** [comment]
> Would his system specs help?

It would definitely help (it was the first thing I asked, but it got diluted). Also, more details. You say you followed the steps for the A5 tutorial, but there's a lot of steps in there, did you do anything unusual? Any decision you had to make that wasn't in the tutorial? Does the error comes right away or how long do you need to wait to see it?

**Comment by manuelma** [comment]
Also, here's another test. Copy/paste the below code into a python script and run it with `python script.py`.

```python
import os
import subprocess
from spinedb_api.spine_db_server import start_spine_db_server, shutdown_spine_db_server
from spinedb_api import DatabaseMapping, import_data

db_url = "sqlite:///spinedb.sqlite"
db_map = DatabaseMapping(db_url, create=True)
import_data(db_map, object_classes=["dog"], objects=[("dog", "brian"), ("dog", "scooby")])
db_map.commit_session("arf!")
server_url = start_spine_db_server(db_url)

code = f'using SpineInterface; using_spinedb("{server_url}"); println(dog())'
subprocess.run(["julia", "-e", code])
os.system(f"julia -e '{code}'")
shutdown_spine_db_server(server_url)
```

**Comment by manuelma** [comment]
> It's a pity that @marcjamali is completely dead in the water

I'm trying to rescue him, but it might take time...

**Comment by manuelma** [comment]
> Ok, we can try that... but what about the threading line of inquiry - how do we test that?

That's exaclty it, the `sleep(1)` could be the blocking call that the server thread needs in order to advance and get the server ready in time.

**Comment by marcjamali** [comment]
Good news.  It works!!! How?  I'm not sure (that's the bad news).

I created a new virtual machine and wanted to reinstall everything and make sure "I" wasn't the problem.

There are only 2 things that I've noticed that were different from the past installations:
- When installing Julia, I noticed a little text saying "Restart Installer as Administrator to install Julia system wide".  I did that although the past installations were made within my Windows user account being an administrator account and the installer worked even without doing that. 
- Second difference is the Spine-Toolbox.  I installed it fetching everything from Github.  Everything as usual except when opening the "File\Settings\Tools" to specify where Julia is.  The interface has changed a little bit but after entering the path for Julia, I had to go through a little automated process that installed the "iJulia" package.

What is this iJulia package?  Is it the Julia installation that changed everything?  Is there a new version of the Spine-Toolbox since a couple of days?

I still have a computer (not a Virtual machine) with the old problem.  If necessary for debugging, I have time next week to work on that (we can meet online with screen sharing or something).

**Comment by manuelma** [comment]
Hi @marcjamali 

Thanks very much for the news. Yes, we changed the settings recently and the fact that you installed IJulia tells something.

To explain, there are two ways of running Julia tools withing Spine Toolbox:
1) In a subprocess
2) In the embedded julia console

In File->Settings->Tools->Julia you have a radio button where you can choose between (1) and (2). If you choose (2) and then press Ok, Spine Toolbox installs IJulia for you. IJulia is the package that enables jupyter kernels in Julia (I'm not sure I can explain all this, but you can google search jupyter to find out more).

My assumption is that you are now using (2), is that correct? Eventually, if you go to File->Settings->Tools->Julia and choose (1), you may encounter the problem in this issue again.

**Comment by manuelma** [comment]
> I have time next week to work on that (we can meet online with screen sharing or something).

This sounds like a very good idea!

**Comment by manuelma** [comment]
I believe what solved this issue was using Julia 1.5. I am positive that it happens with Julia 1.6. I haven't tried 1.2 yet.

The issue also happens only on windows. I was able to 'fix' it by changing the `spinedb_api` code, replacing "localhost" to "127.0.0.1" in the server url. For some reason, it looks like some julia versions don't know or don't want to 'interpret' localhost on windows.
